app/component/itinerary/navigator/hooks/useRealtimeLegs.js

```diff
@@ -97,26 +101,52 @@ function getLegsOfInterest(legs, time) {
     };
   }
 
-  const firstLeg = legs[0];
-  const lastLeg = legs[legs.length - 1];
-  const nextLeg = legs.find(({ start }) => legTime(start) > time);
-  const previousLeg = legs.findLast(({ end }) => legTime(end) < time);
-  const currentLeg = legs.find(
+  const legs = initialLegs.reduce((acc, curr, i, arr) => {
+    acc.push(curr);
+    const next = arr[i + 1];
+
+    // A wait leg is added, if next leg exists but it does not start when current ends
+    if (next && legTime(curr.end) !== legTime(next.start)) {
+      acc.push({ mode: LegMode.Wait, start: curr.end, end: next.start });
+    }
+
+    return acc;
+  }, []);
+
+  const nextLegIdx = legs.findIndex(({ start }) => legTime(start) > time);
+  let currentLeg = legs.find(
     ({ start, end }) => legTime(start) <= time && legTime(end) >= time,
   );
+  let previousLeg = legs.findLast(({ end }) => legTime(end) < time);
+  let nextLeg = legs[nextLegIdx];
+
+  // Indices are shifted by one if a previously completed leg reappears as current.
+  if (
+    nextLeg &&
```

Comment:
I guess flapping should be prevented even at the end of the last leg. In that case just return nextLeg = undefined.

Reply:
It definitely should.

Reply:
Fixed in https://github.com/HSLdevcom/digitransit-ui/pull/5201/commits/0d1563ddf8cefece528f4dacf4e2780e2579beb2